psh/ping/ping.c

```diff
@@ -221,12 +341,54 @@ static int ping_echoreply(int fd, uint8_t *req, uint8_t *resp)
 	icmphdr = (struct icmphdr *)(resp + sizeof(struct iphdr));
 
 	printf("%d bytes received from %s: ttl=%u icmp_seq=%u time=%s\n",
-		bytes, addrstr, iphdr->ttl, ntohs(icmphdr->un.echo.sequence), timestr);
+			bytes, addrstr, iphdr->ttl, ntohs(icmphdr->un.echo.sequence), timestr);
 
 	return 0;
 }
 
 
+#ifdef PSH_IPV6_SUPPORT
+static int ping_echoreply6(int fd, uint8_t *req, uint8_t *resp)
+{
+	time_t sent, elapsed;
+	struct timespec ts;
+	struct icmp6_hdr *icmp6hdr;
+	char addrstr[INET6_ADDRSTRLEN];
+	char timestr[32];
+	int bytes, ret;
+
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	sent = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
+
+	ret = ping_echo6(fd, req, ping_common.reqsz);
+	if (ret < 0) {
+		fprintf(stderr, "ping: Failed to send ECHO REQUEST!\n");
+		return -1;
+	}
+
+	bzero(resp, ping_common.respsz);
+	bytes = ping_reply6(fd, resp, ping_common.respsz, addrstr, sizeof(addrstr));
+	if (bytes < 0) {
+		return -1;
+	}
+
+	clock_gettime(CLOCK_MONOTONIC, &ts);
+	elapsed = ts.tv_sec * 1000000 + ts.tv_nsec / 1000 - sent;
```

Comment:
To avoid overflows, it’s safer to store two timestamps and calculate the elapsed time as:
`(ts2.tv_sec - ts1.tv_sec) * 1000000 + (ts2.tv_nsec - ts1.tv_nsec) / 1000`